dd-java-agent/agent-crashtracking/src/main/java/com/datadog/crashtracking/CrashUploader.java

```diff
@@ -0,0 +1,327 @@
+    // PID can be null if we cannot find it out from the system
+    if (PidHelper.PID != null) {
+      tagsMap.put(PidHelper.PID_TAG, PidHelper.PID.toString());
+    }
+    // Comma separated tags string for V2.4 format
+    tags = String.join(",", tagsToList(tagsMap));
+
+    final Duration requestTimeout =
+        Duration.ofSeconds(
+            configProvider.getInteger(
+                CRASH_TRACKING_UPLOAD_TIMEOUT, CRASH_TRACKING_UPLOAD_TIMEOUT_DEFAULT));
+
+    final OkHttpClient.Builder clientBuilder =
+        new OkHttpClient.Builder()
+            .retryOnConnectionFailure(true)
+            .connectTimeout(requestTimeout)
+            .writeTimeout(requestTimeout)
+            .readTimeout(requestTimeout)
+            .callTimeout(requestTimeout)
+            .proxySelector(AgentProxySelector.INSTANCE);
+
+    final String apmSocketPath = discoverApmSocket(config);
+    if (apmSocketPath != null) {
+      clientBuilder.socketFactory(new UnixDomainSocketFactory(new File(apmSocketPath)));
+    } else if (config.getAgentNamedPipe() != null) {
+      clientBuilder.socketFactory(new NamedPipeSocketFactory(config.getAgentNamedPipe()));
+    }
+
+    if (url.startsWith("http://")) {
+      // force clear text when using http to avoid failures for JVMs without TLS
+      // see: https://github.com/DataDog/dd-trace-java/pull/1582
+      clientBuilder.connectionSpecs(Collections.singletonList(ConnectionSpec.CLEARTEXT));
+    }
+
+    if (configProvider.getString(CRASH_TRACKING_PROXY_HOST) != null) {
+      final Proxy proxy =
+          new Proxy(
+              Proxy.Type.HTTP,
+              new InetSocketAddress(
+                  configProvider.getString(CRASH_TRACKING_PROXY_HOST),
+                  configProvider.getInteger(CRASH_TRACKING_PROXY_PORT)));
+      clientBuilder.proxy(proxy);
+      if (configProvider.getString(CRASH_TRACKING_PROXY_USERNAME) != null) {
+        // Empty password by default
+        final String password =
+            configProvider.getString(CRASH_TRACKING_PROXY_PASSWORD) == null
+                ? ""
+                : configProvider.getString(CRASH_TRACKING_PROXY_PASSWORD);
+        clientBuilder.proxyAuthenticator(
+            (route, response) -> {
+              final String credential =
+                  Credentials.basic(
+                      configProvider.getString(CRASH_TRACKING_PROXY_USERNAME), password);
+              return response
+                  .request()
+                  .newBuilder()
+                  .header("Proxy-Authorization", credential)
+                  .build();
+            });
+      }
+    }
+
+    client = clientBuilder.build();
+  }
+
+  private List<String> tagsToList(final Map<String, String> tags) {
+    return tags.entrySet().stream()
+        .filter(e -> e.getValue() != null && !e.getValue().isEmpty())
+        .map(e -> e.getKey() + ":" + e.getValue())
+        .collect(Collectors.toList());
+  }
```

Comment:
See my previous comment - this may become:
```suggestion
  private String tagsToString(final Map<String, String> tags) {
    return tags.entrySet().stream()
        .filter(e -> e.getValue() != null && !e.getValue().isEmpty())
        .map(e -> e.getKey() + ":" + e.getValue())
        .collect(Collectors.joining(",");
  }
```